storage is not initialized for etched contracts

**Describe the bug**
We get an internal `KeyError` when trying to access storage from a vm.etch'ed contract

Reproduce with:

```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.15;

import "forge-std/Test.sol";
import {console2} from "forge-std/console2.sol";

import {SymTest} from "halmos-cheatcodes/SymTest.sol";

contract NotEtchable {
    address owner;

    constructor() {
        owner = msg.sender;
    }

    function beepBoop() public {
        require(msg.sender == owner, "NotEtchable: only owner can beep boop");
        console2.log("beep boop");
    }
}

contract Test95 is Test, SymTest {
    NotEtchable target;

    function setUp() public {
        // etch does not run the constructor, so owner is never set
        vm.etch(address(42), type(NotEtchable).runtimeCode);

        target = NotEtchable(address(42));
    }

    function test_beep_boop() external {
        target.beepBoop();
    }
}
```

```bash
halmos --function test_beep_boop

Executing test_beep_boop
[ERROR] test_beep_boop()
KeyError: 42
Traceback (most recent call last):
  File "/Users/karma/projects/halmos/src/halmos/__main__.py", line 950, in run_sequential
    test_result = run(setup_ex, run_args.abi, fun_info, test_config)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/karma/projects/halmos/src/halmos/__main__.py", line 666, in run
    for idx, ex in enumerate(exs):
  File "/Users/karma/projects/halmos/src/halmos/sevm.py", line 2829, in run
    ex.st.push(self.sload(ex, ex.this(), slot))
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/karma/projects/halmos/src/halmos/sevm.py", line 1775, in sload
    return self.storage_model.load(ex, addr, loc)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/karma/projects/halmos/src/halmos/sevm.py", line 1382, in load
    cls.init(ex, addr, loc, size_keys)
  File "/Users/karma/projects/halmos/src/halmos/sevm.py", line 1372, in init
    mapping = ex.storage[addr].mapping
              ~~~~~~~~~~^^^^^^
KeyError: 42
Symbolic test result: 0 passed; 1 failed; time: 0.06s

[time] total: 1.32s (build: 0.71s, load: 0.54s, tests: 0.07s)
```

--- a/.gitignore
+++ b/.gitignore
@@ -14,7 +14,7 @@ __pycache__/
 
 # Environments
 .env
-.venv
+.venv*
 env/
 venv/
 

--- a/src/halmos/cheatcodes.py
+++ b/src/halmos/cheatcodes.py
@@ -652,13 +652,11 @@ class hevm_cheat_code:
                 ex, store_account, stack, step_id, branching=False
             )
 
-            if store_account_alias is not None:
-                store_account = store_account_alias
+            if store_account_alias is None:
+                error_msg = f"vm.store() is not allowed for a nonexistent account: {hexify(store_account)}"
+                raise HalmosException(error_msg)
 
-            # vm.store() initializes storage if the account is not already initialized
-            ex.storage.setdefault(store_account, sevm.mk_storagedata())
-
-            sevm.sstore(ex, store_account, store_slot, store_value)
+            sevm.sstore(ex, store_account_alias, store_slot, store_value)
             return ret
 
         # vm.load(address,bytes32)

--- a/src/halmos/cheatcodes.py
+++ b/src/halmos/cheatcodes.py
@@ -652,11 +652,13 @@ class hevm_cheat_code:
                 ex, store_account, stack, step_id, branching=False
             )
 
-            if store_account_alias is None:
-                error_msg = f"vm.store() is not allowed for a nonexistent account: {hexify(store_account)}"
-                raise HalmosException(error_msg)
+            if store_account_alias is not None:
+                store_account = store_account_alias
 
-            sevm.sstore(ex, store_account_alias, store_slot, store_value)
+            # vm.store() initializes storage if the account is not already initialized
+            ex.storage.setdefault(store_account, sevm.mk_storagedata())
+
+            sevm.sstore(ex, store_account, store_slot, store_value)
             return ret
 
         # vm.load(address,bytes32)
@@ -720,6 +722,9 @@ class hevm_cheat_code:
             code_loc = 4 + code_offset + 32
             code_bytes = arg[code_loc : code_loc + code_length]
             ex.set_code(who, code_bytes)
+
+            # vm.etch() initializes but does not clear storage
+            ex.storage.setdefault(who, sevm.mk_storagedata())
 
             return ret
 

--- a/src/halmos/sevm.py
+++ b/src/halmos/sevm.py
@@ -2252,6 +2252,7 @@ class SEVM:
 
         # setup new account
         ex.set_code(new_addr, Contract(b""))  # existing code must be empty
+
         # existing storage may not be empty and reset here
         ex.storage[new_addr] = self.mk_storagedata()
 

--- a/tests/expected/all.json
+++ b/tests/expected/all.json
@@ -1109,6 +1109,26 @@
                 "num_bounded_loops": null
             }
         ],
+        "test/Foundry.t.sol:TestNotEtchFriendly": [
+            {
+                "name": "check_etch_no_owner(address)",
+                "exitcode": 0,
+                "num_models": 0,
+                "models": null,
+                "num_paths": null,
+                "time": null,
+                "num_bounded_loops": null
+            },
+            {
+                "name": "check_etch_then_store()",
+                "exitcode": 0,
+                "num_models": 0,
+                "models": null,
+                "num_paths": null,
+                "time": null,
+                "num_bounded_loops": null
+            }
+        ],
         "test/Getter.t.sol:GetterTest": [
             {
                 "name": "check_Getter(uint256)",

--- a/tests/regression/test/Foundry.t.sol
+++ b/tests/regression/test/Foundry.t.sol
@@ -124,3 +124,47 @@ contract FoundryTest is Test {
     //     assertEq(code, who.code);
     // }
 }
+
+
+contract NotEtchFriendly {
+    address owner;
+
+    constructor() {
+        owner = msg.sender;
+    }
+
+    function beepBoop() public {
+        console2.log("owner is", owner);
+        require(msg.sender == owner, "NotEtchFriendly: only owner can beep boop");
+    }
+}
+
+contract TestNotEtchFriendly is Test {
+    NotEtchFriendly target;
+
+    function setUp() public {
+        /// @dev this is supported in foundry, but not halmos (can't vm.store to uninitialized account)
+        // make address(this) the owner of the yet-to-be-deployed contract
+        // vm.store(address(42), 0, bytes32(uint256(uint160(address(this)))));
+
+        // etch does not run the constructor, so owner is not set by the constructor
+        // additionally, vm.etch does not reset storage (unlike CREATE2)
+        vm.etch(address(42), type(NotEtchFriendly).runtimeCode);
+
+        target = NotEtchFriendly(address(42));
+    }
+
+    function check_etch_no_owner(address sender) external {
+        vm.prank(sender);
+        target.beepBoop();
+
+        assertEq(sender, address(0));
+    }
+
+    function check_etch_then_store() external {
+        // make address(this) the owner of the contract, emulating the constructor that did not run
+        vm.store(address(42), 0, bytes32(uint256(uint160(address(this)))));
+
+        target.beepBoop();
+    }
+}

